.github/workflows/justtest.yaml

```diff
@@ -0,0 +1,699 @@
+# The idea of this workflow is to be simple and developer-friendly.
+#
+# You don't have to read all the comments: it is better to just
+# follow the code. The workflow has a lot of remarks regarding
+# a motivation to use this and not that. They're not necessary to
+# understand what is going on.
+#
+# {{{ Design choices
+#
+# Overview
+# ========
+#
+# To achieve the simplicity several choices are made:
+#
+# * Use GitHub hosted runners.
+# * Don't attempt to enhance GitHub Actions.
+# * No secrets or deployments.
+# * No hacks for unusual tests.
+# * Keep all the logic within a single file.
+#
+# However, there are several choices, where a complication is
+# taken as acceptable in the sense that the workflow is still
+# readable without a deep context.
+#
+# * Perform several builds in one workflow.
+# * Split testing to jobs per test suite.
+# * Cache submodules build.
+#
+# Let's look into these points.
+#
+# Self-constraints
+# ================
+#
+# Use GitHub hosted runners
+# -------------------------
+#
+# This way we don't deal with any infrastructural questions here.
+#
+# For example, we don't have to use docker and support our own
+# docker images for testing.
+#
+# Also, actions/checkout needs a couple of hacks to make it
+# working on a self-hosted runner (at least git clean for
+# submodules, fetching tags for submodules, removing lock files
+# left by a previous job).
+#
+# At the same time the GitHub hosted runners provide a good
+# parallelization level (60 jobs in parallel on the Team plan)
+# and we shouldn't experience large delays during development.
```

Comment:
Fun fact: we have 36 always-online amd64 runners enabled for this repository. So it seems like we have even less capacity with our own infrastucture. The main point for enabling our own infra was to don't wait for CI too long (but ~20 min seems long for me).